_posts/2017-03-05-artsy-technology-stack-2017.md

```diff
@@ -0,0 +1,176 @@
+
+# Search
+
+We completed our full migration from [Solr](http://lucene.apache.org/solr/) to [Elasticsearch](https://www.elastic.co/) in the last 18 months, and now use the latter across artsy.net, from all our artwork filter interfaces through to our real-time artwork similarity feature. Elasticsearch gives us high availability clustering features out of the box and easy horizontal scaling.
+
+# Platform Services
+
+As Artsy's business has grown more complex, so has the data and concepts handled by its core API. We've begun supporting certain product areas with separate, dedicated API services, and even extracting existing API domains into separate services when practical. These services tend to expose simple [REST](https://en.wikipedia.org/wiki/Representational_state_transfer)-ful HTTP APIs, maintain separate data sources, and even do their own [authentication](/blog/2016/10/26/jwt-artsy-journey/). This has certain advantages:
+
+* Each system can be deployed and scaled independently.
+* Each chooses the best-suited languages and technologies for its purpose.
+* Code bases remain more focused and developers' cognitive overhead is minimized.
+
+Balancing these out are some very real disadvantages:
+
+* Development must sometimes touch multiple systems.
+* Some data is copied between services. These can become out-of-sync, though we always try to have a single _authoritative_ source in such cases.
+* Deploys must be coordinated.
+
+At our size and complexity, a single code base is simply impractical. So, we've tried to be consistent in the coding, deployment, monitoring, and logging practices of these services. The more repeatable and disciplined our process, the less overhead is introduced by additional systems.
+
+We've also explored alternate communication patterns, so systems aren't as dependent on each other's APIs. Recently we've begun publishing a stream of interesting data events from our core systems. Other systems can simply subscribe to the notifications they care about, so the source system doesn't need to be concerned about integrating with one more destination. After experimenting with [Kafka](https://kafka.apache.org/) but finding it hard to manage, we switched to [RabbitMQ](https://www.rabbitmq.com/) for this purpose.
+
+## Provisioning
+
+We use an open source in-house [Docker](https://www.docker.com/) workflow toolkit called [Hokusai](https://github.com/artsy/hokusai) to manage containerised deployment via a [Kubernetes](https://kubernetes.io/) cluster. All our new infrastructure is configured in code using [Terraform](https://www.terraform.io/). This new workflow is reducing our dependence on Heroku, giving us more flexibility in our deployments and a more efficient use of server resources.
```

Comment:
Two edits:
- A first sentence on Terraform and the effects of its adoption
- "We started developing..." as Hokusai is still yet to be a v1 release
- "...in order to manage a containerised workflow, CI and deployment to Kubernetes" as Hokusai aims to cover the whole shebang - dev / test / CI / artifacts and deployment
```
All our new AWS infrastructure is configured in code using [Terraform](https://www.terraform.io/). This approach has allowed us to quickly replicate entire deployments along with their dependencies and has increased visibility into the state of our infrastructure across our teams.  We started developing an open source [Docker](https://www.docker.com/) workflow toolkit called [Hokusai](https://github.com/artsy/hokusai) in order to manage a containerized workflow, CI and deployment to [Kubernetes](https://kubernetes.io/). Our Kubernetes clusters are managed using [Kops](https://github.com/kubernetes/kops) and similarly provisioned using Terraform. This new workflow is reducing our dependence on Heroku, giving us more flexibility in our deployments and a more efficient use of server resources.
```